packages/run-protocol/src/interchainPool.js

```diff
@@ -0,0 +1,138 @@
+// @ts-check
+import { E, Far } from '@endo/far';
+import { AmountMath, AssetKind, makeIssuerKit } from '@agoric/ertp';
+import { offerTo } from '@agoric/zoe/src/contractSupport/index.js';
+
+const { details: X, quote: q } = assert;
+
+const COSMOS_DECIMALS = 6;
+
+/**
+ * Given sufficient IST, create an issuer for an IBC denom
+ * and an invitation to add a pool for it to the AMM.
+ *
+ * @param {ZCF<InterchainPoolTerms>} zcf
+ * @param {{
+ *   bankManager: ERef<BankManager>,
+ * }} pa
+ *
+ * @typedef {{
+ *   minimumCentral: Amount<'nat'>,
+ *   amm: ERef<XYKAMMPublicFacet>,
+ * }} InterchainPoolTerms
+ *
+ * @typedef {{
+ *   denom: string,
+ *   decimalPlaces: number,
+ * }} AssetDetail
+ */
+export const start = (zcf, { bankManager }) => {
+  const {
+    brands: { Central: centralBrand },
+    minimumCentral,
+    amm,
+  } = zcf.getTerms();
+  AmountMath.coerce(centralBrand, minimumCentral);
+
+  let kwNonce = 0;
+
+  /**
+   * @param {ZCFSeat} seat
+   * @param {AssetDetail} detail
+   */
+  const step1Handler = async (seat, detail) => {
+    assert.typeof(
+      detail,
+      'object',
+      'offer arg must be { denom, decimalPlaces? }',
+    );
+    const { denom, decimalPlaces = COSMOS_DECIMALS } = detail;
+    assert.typeof(denom, 'string');
+    assert.typeof(decimalPlaces, 'number');
+
+    const {
+      give: { Central: centralAmt },
+    } = seat.getProposal();
+    assert(
+      AmountMath.isGTE(centralAmt, minimumCentral),
+      X`at least ${q(minimumCentral)} required; only ${q(centralAmt)} given`,
+    );
+
+    const interAsset = makeIssuerKit(
+      denom,
+      AssetKind.NAT,
+      harden({ decimalPlaces }),
+    );
+    kwNonce += 1;
+
+    const liquidityIssuer = await E(amm).addPool(
+      interAsset.issuer,
+      `Interchain${kwNonce}`,
+    );
+    await Promise.all([
+      zcf.saveIssuer(interAsset.issuer, `Interchain${kwNonce}`),
+      zcf.saveIssuer(liquidityIssuer, `Liquidity${kwNonce}`),
+    ]);
+
+    /** @type { OfferHandler } */
+    const step2Handler = async seat2 => {
+      const {
+        give: { Secondary: secondaryAmt },
+        want: { Liquidity: liqAmt },
+      } = seat2.getProposal();
+      AmountMath.coerce(interAsset.brand, secondaryAmt);
+      const liqBrand = await E(liquidityIssuer).getBrand();
+      AmountMath.coerce(liqBrand, liqAmt);
+
+      const proposal = harden({
+        give: {
+          Central: centralAmt,
+          Secondary: secondaryAmt,
+        },
+        want: { Liquidity: liqAmt },
+      });
+
+      seat2.incrementBy(seat.decrementBy(harden({ Central: centralAmt })));
+      zcf.reallocate(seat, seat2);
+
+      const invitation = await E(amm).makeAddLiquidityInvitation();
+      const { userSeatPromise, deposited } = await offerTo(
+        zcf,
+        invitation,
+        undefined,
+        proposal,
+        seat2,
+      );
+      return deposited.then(_ => {
+        seat.exit();
+        seat2.exit();
```

Comment:
I think there should be a `.catch()` clause here that returns the money from seat2 to seat1 and exits the seat.

I would test it by setting the AMM's minimum higher than the minimum enforced here.

Reply:
I plan to get the AMM's minimum on demand, so that test wouldn't last.

What is it that could throw? `seat.exit()` or `deposited`? In either case, what difference does it make which seat the money is returned in?

Reply:
> what difference does it make which seat the money is returned in?

I thought seat2 was purely internal, but I now see that I was wrong.

It still seems surprising that the caller has to check both seats to see if there's money there.  

If someone has already saved an issuer on the AMM and called it 'Interchain5', then the `saveIssuer()` call will fail, and the allocation will be stuck on seat2.

Reply:
> It still seems surprising that the caller has to check both seats to see if there's money there.

In discussion with @dtribble and co., we didn't see an obvious improvement. (There are 2 offers, so having 2 seats to check on failure never seemed odd, to me.)

> If someone has already saved an issuer on the AMM and called it 'Interchain5', then the `saveIssuer()` call will fail

The `saveIssuer()` call here doesn't depend on issuers in the AMM, as we discussed. (There could be a collision at `addPool`, but that's before the `saveIssuer()` here.)